modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java

```diff
@@ -0,0 +1,2728 @@
+import org.elasticsearch.painless.symbol.Decorations.StandardPainlessConstructor;
+import org.elasticsearch.painless.symbol.Decorations.StandardPainlessField;
+import org.elasticsearch.painless.symbol.Decorations.StandardPainlessInstanceBinding;
+import org.elasticsearch.painless.symbol.Decorations.StandardPainlessMethod;
+import org.elasticsearch.painless.symbol.Decorations.StaticType;
+import org.elasticsearch.painless.symbol.Decorations.TargetType;
+import org.elasticsearch.painless.symbol.Decorations.TypeParameters;
+import org.elasticsearch.painless.symbol.Decorations.UnaryType;
+import org.elasticsearch.painless.symbol.Decorations.UpcastPainlessCast;
+import org.elasticsearch.painless.symbol.Decorations.ValueType;
+import org.elasticsearch.painless.symbol.Decorations.Write;
+import org.elasticsearch.painless.symbol.FunctionTable;
+import org.elasticsearch.painless.symbol.FunctionTable.LocalFunction;
+import org.elasticsearch.painless.symbol.ScriptScope;
+import org.elasticsearch.painless.symbol.SemanticScope;
+import org.elasticsearch.painless.symbol.SemanticScope.FunctionScope;
+import org.elasticsearch.painless.symbol.SemanticScope.LambdaScope;
+import org.elasticsearch.painless.symbol.SemanticScope.Variable;
+
+import java.lang.reflect.Modifier;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+import static org.elasticsearch.painless.lookup.PainlessLookupUtility.typeToCanonicalTypeName;
+import static org.elasticsearch.painless.symbol.SemanticScope.newFunctionScope;
+
+public class DefaultSemanticAnalysisPhase extends UserTreeBaseVisitor<SemanticScope> {
+
+    public void decorateWithCast(AExpression userExpressionNode, SemanticScope semanticScope) {
+        Location location = userExpressionNode.getLocation();
+        Class<?> valueType = semanticScope.getDecoration(userExpressionNode, ValueType.class).getValueType();
+        Class<?> targetType = semanticScope.getDecoration(userExpressionNode, TargetType.class).getTargetType();
+        boolean isExplicitCast = semanticScope.getCondition(userExpressionNode, Explicit.class);
+        boolean isInternalCast = semanticScope.getCondition(userExpressionNode, Internal.class);
+
+        PainlessCast painlessCast = AnalyzerCaster.getLegalCast(location, valueType, targetType, isExplicitCast, isInternalCast);
+
+        if (painlessCast != null) {
+            semanticScope.putDecoration(userExpressionNode, new ExpressionPainlessCast(painlessCast));
+        }
+    }
+
+    public void visit(ANode userNode, SemanticScope semanticScope) {
+        if (userNode != null) {
+            userNode.visit(this, semanticScope);
+        }
+    }
+
+    public void checkedVisit(AExpression userExpressionNode, SemanticScope semanticScope) {
```

Comment:
Provide a one sentence explanation for each public method.

Reply:
Will do. This is going to take a bit of time.